server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -0,0 +1,424 @@
+remoteSchemaInterface
+  :: forall n m
+  . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> G.InterfaceTypeDefinition [G.Name]
+  -> m (Parser 'Output n ())
+remoteSchemaInterface schemaDoc defn@(G.InterfaceTypeDefinition description name _directives fields possibleTypes) =
+  P.memoizeOn 'remoteSchemaObject defn do
+  subFieldParsers <- traverse (remoteField' schemaDoc) fields
+  objs :: [Parser 'Output n ()] <-
+    traverse (getObject >=> remoteSchemaObject schemaDoc) possibleTypes
+  -- In the Draft GraphQL spec (> June 2018), interfaces can themselves
+  -- implement superinterfaces.  In the future, we may need to support this
+  -- here.
+  when (null subFieldParsers) $
+    throw400 RemoteSchemaError $ "List of fields cannot be empty for interface " <> squote name
+  -- TODO: another way to obtain 'possibleTypes' is to lookup all the object
+  -- types in the schema document that claim to implement this interface.  We
+  -- should have a check that expresses that that collection of objects is equal
+  -- to 'possibelTypes'.
+  pure $ void $ P.selectionSetInterface name description subFieldParsers objs
+  where
+    getObject :: G.Name -> m G.ObjectTypeDefinition
+    getObject objectName =
+      onNothing (lookupObject schemaDoc objectName) $
+        case lookupInterface schemaDoc objectName of
+          Nothing -> throw400 RemoteSchemaError $ "Could not find type " <> squote objectName
+            <> ", which is defined as a member type of Union " <> squote name
+          Just _  -> throw400 RemoteSchemaError $ "Union type " <> squote name <>
+            " can only include object types. It cannot include " <> squote objectName
+
+-- | 'remoteSchemaUnion' returns a output parser for a given 'UnionTypeDefinition'.
+remoteSchemaUnion
+  :: forall n m
+  . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> G.UnionTypeDefinition
+  -> m (Parser 'Output n ())
+remoteSchemaUnion schemaDoc defn@(G.UnionTypeDefinition description name _directives objectNames) =
+  P.memoizeOn 'remoteSchemaObject defn do
+  objDefs <- traverse getObject objectNames
+  objs :: [Parser 'Output n ()] <- traverse (remoteSchemaObject schemaDoc) objDefs
+  when (null objs) $
+    throw400 RemoteSchemaError $ "List of member types cannot be empty for union type " <> squote name
+  pure $ void $ P.selectionSetUnion name description objs
+  where
+    getObject :: G.Name -> m G.ObjectTypeDefinition
+    getObject objectName =
+      onNothing (lookupObject schemaDoc objectName) $
+        case lookupInterface schemaDoc objectName of
+          Nothing -> throw400 RemoteSchemaError $ "Could not find type " <> squote objectName
+            <> ", which is defined as a member type of Union " <> squote name
+          Just _  -> throw400 RemoteSchemaError $ "Union type " <> squote name <>
+            " can only include object types. It cannot include " <> squote objectName
```

Comment:
AFAICT, this is exactly the same function as the previous `getObject`. Maybe create a separate function that they can both call?

Reply:
Actually there was supposed to be a small difference between them: unions vs interfaces in the message. But the current version got this wrong. So I generalized the function, combining the two variants.
When the current draft GraphQL spec is released, we may need to re-split them, as membership of unions may be defined differently from membership of interfaces.